core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -422,11 +425,17 @@ private[log] class Cleaner(val id: Int,
       for (entry <- messages.shallowIterator) {
         val size = MessageSet.entrySize(entry.message)
         stats.readMessage(size)
-        if (entry.message.compressionCodec == NoCompressionCodec) {
+        val sourceCodec: CompressionCodec = entry.message.compressionCodec
+        val targetCodec = BrokerCompressionCodec.getTargetCompressionCodec(brokerCompressionType, sourceCodec)
+        if (sourceCodec == NoCompressionCodec) {
           if (shouldRetainMessage(source, map, retainDeletes, entry)) {
             val convertedMessage = entry.message.toFormatVersion(messageFormatVersion)
-            ByteBufferMessageSet.writeMessage(writeBuffer, convertedMessage, entry.offset)
-            stats.recopyMessage(size)
+            if (targetCodec == NoCompressionCodec) { // same as sourceCodec
+              ByteBufferMessageSet.writeMessage(writeBuffer, convertedMessage, entry.offset)
+              stats.recopyMessage(size)
+            } else {
+              compressMessages(writeBuffer, targetCodec, messageFormatVersion, Seq(new MessageAndOffset(convertedMessage, entry.offset)))
```

Comment:
Yes, this can be tricky. One way is to keep the original compression codec if changing the codec makes the message oversized. However, in theory, it's possible that after we take a message out from a set and recompresses the set with the same codec, max.message.bytes is exceeded. Not sure quite what to do in this case. Perhaps, we can take the message and just deal with the oversized message as in KAFKA-3810?